chips/stm32f4xx/src/flash.rs

```diff
@@ -0,0 +1,711 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright OxidOS Automotive SRL.
+
+#![deny(missing_docs)]
+//! STM32F4xx flash driver
+//!
+//! This driver provides basic functionalities for the entire STM32F4 series.
+//!
+//! # Features
+//!
+//! - [x] Configuring latency based on the system clock frequency
+//!
+//! # Missing features
+//!
+//! - [ ] Support for different power supplies
+//! - [ ] Instruction prefetch
+//! - [ ] Instruction and data cache
+//!
+//!
+//! # Usage
+//!
+//! To use this driver, a reference to the Flash peripheral is required:
+//!
+//! ```rust,ignore
+//! // Inside the board main.rs
+//! let flash = &peripherals.stm32f4.flash;
+//! ```
+//!
+//! ## Retrieve the current flash latency
+//!
+//! ```rust,ignore
+//! let flash_latency = flash.get_latency() as usize;
+//! debug!("Current flash latency is {}", flash_latency);
+//! ```
+
+use kernel::debug;
+use kernel::utilities::registers::interfaces::{ReadWriteable, Readable};
+use kernel::utilities::registers::{register_bitfields, register_structs, ReadWrite, WriteOnly};
+use kernel::utilities::StaticRef;
+use kernel::ErrorCode;
+
+register_structs! {
+    /// FLASH
+    FlashRegisters {
+        /// Flash access control register
+        (0x000 => acr: ReadWrite<u32, ACR::Register>),
+        /// Flash key register
+        (0x004 => keyr: WriteOnly<u32>),
+        /// Flash option key register
+        (0x008 => optkeyr: WriteOnly<u32>),
+        /// Status register
+        (0x00C => sr: ReadWrite<u32, SR::Register>),
+        /// Control register
+        (0x010 => cr: ReadWrite<u32, CR::Register>),
+        /// Flash option control register
+        (0x014 => optcr: ReadWrite<u32, OPTCR::Register>),
+        /// Flash option control register 1
+        // NOTE: This register is present only on some chip models
+        (0x018 => optcr1: ReadWrite<u32>),
+        (0x01C => @END),
+    }
+}
+
+register_bitfields![u32,
```

Comment:
This is the common superset of all possible bitfieds for all chips. This was done to avoid code duplication.